File: mifosng-android/src/main/java/com/mifos/objects/PaymentTypeOption.java
/*
 * This project is licensed under the open source MPL V2.
 * See https://github.com/openMF/android-client/blob/master/LICENSE.md
 */

package com.mifos.objects;

import android.os.Parcel;
import android.os.Parcelable;

import com.google.gson.annotations.SerializedName;
import com.mifos.api.local.MifosBaseModel;
import com.mifos.api.local.MifosDatabase;
import com.raizlabs.android.dbflow.annotation.Column;
import com.raizlabs.android.dbflow.annotation.ModelContainer;
import com.raizlabs.android.dbflow.annotation.PrimaryKey;
import com.raizlabs.android.dbflow.annotation.Table;

@Table(database = MifosDatabase.class)
@ModelContainer
public class PaymentTypeOption extends MifosBaseModel implements Comparable<PaymentTypeOption>,
        Parcelable {


    @SerializedName("id")
    @PrimaryKey
    Integer id;

    @SerializedName("name")
    @Column
    String name;

    @SerializedName("description")
    @Column
    String description;

    @SerializedName("isCashPayment")
    @Column
    Boolean isCashPayment;

    @SerializedName("position")
    @Column
    Integer position;

    public Integer getId() {
        return id;
    }

    public void setId(Integer id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public Integer getPosition() {
        return position;
    }

    public void setPosition(Integer position) {
        this.position = position;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public Boolean getCashPayment() {
        return isCashPayment;
    }

    public void setCashPayment(Boolean cashPayment) {
        isCashPayment = cashPayment;
    }

    /**
     * Compares two payment type options on the basis
     * of their position specified.
     *
     * @param another
     * @return
     */

    @Override
    public int compareTo(PaymentTypeOption another) {

        if (this.position < another.position) {
            return -1;
        } else if (this.position > another.position) {
            return 1;
        } else {
            return 0;
        }
    }

    @Override
    public String toString() {
        return "PaymentTypeOption{" +
                "id=" + id +
                ", name='" + name + '\'' +
                ", description='" + description + '\'' +
                ", isCashPayment=" + isCashPayment +
                ", position=" + position +
                '}';
    }

    public PaymentTypeOption() {
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeValue(this.id);
        dest.writeString(this.name);
        dest.writeString(this.description);
        dest.writeValue(this.isCashPayment);
        dest.writeValue(this.position);
    }

    protected PaymentTypeOption(Parcel in) {
        this.id = (Integer) in.readValue(Integer.class.getClassLoader());
        this.name = in.readString();
        this.description = in.readString();
        this.isCashPayment = (Boolean) in.readValue(Boolean.class.getClassLoader());
        this.position = (Integer) in.readValue(Integer.class.getClassLoader());
    }

    public static final Creator<PaymentTypeOption> CREATOR = new Creator<PaymentTypeOption>() {
        @Override
        public PaymentTypeOption createFromParcel(Parcel source) {
            return new PaymentTypeOption(source);
        }

        @Override
        public PaymentTypeOption[] newArray(int size) {
            return new PaymentTypeOption[size];
        }
    };
}

File: core/database/src/main/java/com/mifos/core/objects/PaymentTypeOption.kt
/*
 * Copyright 2024 Mifos Initiative
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/.
 *
 * See https://github.com/openMF/android-client/blob/master/LICENSE.md
 */
package com.mifos.core.objects

import android.os.Parcel
import android.os.Parcelable
import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.PrimaryKey
import com.google.gson.annotations.SerializedName
import com.mifos.core.model.MifosBaseModel

@Entity(tableName = "PaymentTypeOption")
data class PaymentTypeOption(
    @SerializedName("id")
    @PrimaryKey
    var id: Int = 0,

    @SerializedName("name")
    @ColumnInfo("name")
    var name: String = "",

    @SerializedName("description")
    @ColumnInfo("description")
    var description: String? = null,

    @SerializedName("isCashPayment")
    @ColumnInfo("isCashPayment")
    var isCashPayment: Boolean? = null,

    @SerializedName("position")
    @ColumnInfo("position")
    var position: Int? = null,
) : MifosBaseModel(), Comparable<PaymentTypeOption>, Parcelable {

    override fun compareTo(another: PaymentTypeOption): Int {
        return position?.compareTo(another.position ?: 0) ?: 0
    }

    override fun toString(): String {
        return "PaymentTypeOption{" +
            "id=$id, " +
            "name='$name', " +
            "description='$description', " +
            "isCashPayment=$isCashPayment, " +
            "position=$position" +
            '}'
    }

    constructor(parcel: Parcel) : this(
        parcel.readValue(Int::class.java.classLoader) as Int,
        parcel.readString() as String,
        parcel.readString(),
        parcel.readValue(Boolean::class.java.classLoader) as? Boolean,
        parcel.readValue(Int::class.java.classLoader) as? Int,
    )

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeValue(id)
        dest.writeString(name)
        dest.writeString(description)
        dest.writeValue(isCashPayment)
        dest.writeValue(position)
    }

    companion object CREATOR : Parcelable.Creator<PaymentTypeOption> {
        override fun createFromParcel(parcel: Parcel): PaymentTypeOption {
            return PaymentTypeOption(parcel)
        }

        override fun newArray(size: Int): Array<PaymentTypeOption?> {
            return arrayOfNulls(size)
        }
    }
}
